tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,151 @@
+GOPRIVATE and are automatically retrieved from the host environment
+when docker-go is envoked.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+
+        // If a module directory was not provided, set as an empty path.
+        // By default, without a provided module directory, tar will be passed
+        // the first directory found in the archieve as the top level Go module
+        let default_empty_path = PathBuf::from("");
+        let mod_dir = gomod.bundle_path.as_ref().unwrap_or(&default_empty_path);
+
+        // Use a default "bundle-{name-of-tar-file}" if no output path was provided
+        let path = path_arg.to_string_lossy();
+        let default_output_path = PathBuf::from(format!("bundled-{}", path));
+        let output_path_arg = gomod
+            .bundle_output_path
+            .as_ref()
+            .unwrap_or(&default_output_path);
+        println!(
+            "cargo:rerun-if-changed={}",
+            output_path_arg.to_string_lossy()
+        );
+
+        // Our SDK and toolchain are picked by the external `cargo make` invocation.
+        let sdk = env::var("BUILDSYS_SDK_IMAGE").context(error::EnvironmentSnafu {
+            var: "BUILDSYS_SDK_IMAGE",
+        })?;
+
+        let args = DockerGoArgs {
+            module_path: package_dir,
+            sdk_image: sdk,
+            go_mod_cache: &root_dir.join(".gomodcache"),
+            command: format!(
+                // This script will inspect the top level directory found in the archive
+                // and use that as the default Go module path
+                // if no explicit module path was provided.
+                // It will then untar the archive, vendor the Go dependencies,
+                // create a new archieve using the {module-path}/vendor directory
+                // and name it the output path provided.
+                // If no output path was given, it defaults to "bundled-{tar-file-name}".
+                // Finally, it cleans up by removing the untar'd source code.
+                // The upstream archive remains and both tar files can be used during packaging.
+                "toplevel=$(tar tf {input} | head -1)
+                tar xf {input} &&
+                pushd $(if [ -z {moddir} ]; then echo $toplevel; else echo {moddir}; fi) &&
```

Comment:
I'd recommend not doing the Dockerfile-ish technique of writing `&&` everywhere if it's not necessary.

This would be more readable as:
```
if [ -n "${moddir}" ] ; then
  targetdir="${moddir}"
else
  targetdir="${toplevel}"
fi

pushd "${targetdir}"
# do the vendoring
popd

if [ ! -d "${targetdir}/vendor ] ;
  echo "no vendor directory found" >&2
  exit 1
fi

tar czf ... "${targetdir}/vendor"

# clean up - why?
rm -rf "${targetdir}"
```